### PR TITLE
[SPARK-20427][SQL] Read JDBC table use custom schema

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

Auto generated Oracle schema some times not we expect:

- `number(1)` auto mapped to BooleanType, some times it's not we expect, per [SPARK-20921](https://issues.apache.org/jira/browse/SPARK-20921).
-  `number` auto mapped to Decimal(38,10), It can't read big data, per [SPARK-20427](https://issues.apache.org/jira/browse/SPARK-20427).

This PR fix this issue by custom schema as follows:
```scala
val props = new Properties()
props.put("customSchema", "ID decimal(38, 0), N1 int, N2 boolean")
val dfRead = spark.read.schema(schema).jdbc(jdbcUrl, "tableWithCustomSchema", props)
dfRead.show()
```
or
```sql
CREATE TEMPORARY VIEW tableWithCustomSchema
USING org.apache.spark.sql.jdbc
OPTIONS (url '$jdbcUrl', dbTable 'tableWithCustomSchema', customSchema'ID decimal(38, 0), N1 int, N2 boolean')
```

## How was this patch tested?

unit tests

### CODE REVIEW

**Comment by wangyum** [comment]
retest please.

**Comment by wangyum** [comment]
Jenkins, retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Metadata.scala [hunk 1]
```diff
@@ -273,6 +273,9 @@ class MetadataBuilder {
   /** Puts a [[Metadata]] array. */
   def putMetadataArray(key: String, value: Array[Metadata]): this.type = put(key, value)
 
+  /** Puts a name. */
+  def putName(name: String): this.type = put("name", name)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Metadata.scala]: This interface change is not desired. See the PR https://github.com/apache/spark/pull/16209

You can further enhance our parser by supporting the data types that are not natively supported by Spark.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 2]
```diff
@@ -907,7 +907,7 @@ class JDBCSuite extends SparkFunSuite
     assert(new JDBCOptions(CaseInsensitiveMap(parameters)).asConnectionProperties.isEmpty)
   }
 
-  test("SPARK-16848: jdbc API throws an exception for user specified schema") {
+  ignore("SPARK-16848: jdbc API throws an exception for user specified schema") {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: ?

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala [hunk 3]
```diff
@@ -198,4 +205,37 @@ class OracleIntegrationSuite extends DockerJDBCIntegrationSuite with SharedSQLCo
...
+      "requirement failed: Decimal precision 39 exceeds max precision 38"))
+
+    // custom schema can read data
+    val schema = StructType(Seq(
+      StructField("ID", DecimalType(DecimalType.MAX_PRECISION, 0), true,
+        new MetadataBuilder().putString("name", "ID").build()),
+      StructField("N1", IntegerType, true, new MetadataBuilder().putString("name", "N1").build()),
+      StructField("N2", BooleanType, true, new MetadataBuilder().putString("name", "N2").build())))
+
+    val dfRead = spark.read.schema(schema).jdbc(jdbcUrl, "custom_column_types", new Properties())
```
Comment (gatorsmile) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: ?

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala [hunk 4]
```diff
@@ -198,4 +205,37 @@ class OracleIntegrationSuite extends DockerJDBCIntegrationSuite with SharedSQLCo
...
+      spark.read.jdbc(jdbcUrl, "custom_column_types", new Properties()).collect()
+    }
+    assert(e.getMessage.contains(
+      "requirement failed: Decimal precision 39 exceeds max precision 38"))
+
+    // custom schema can read data
+    val schema = StructType(Seq(
+      StructField("ID", DecimalType(DecimalType.MAX_PRECISION, 0), true,
+        new MetadataBuilder().putString("name", "ID").build()),
+      StructField("N1", IntegerType, true, new MetadataBuilder().putString("name", "N1").build()),
```
Comment (gatorsmile) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: Why adding `new MetadataBuilder().putString("name", "N1").build()`?

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 2 above)
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: JDBC didn't support specified schema before:
https://github.com/apache/spark/blob/v2.2.0-rc5/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L188

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala (hunk 4 above)
Comment (wangyum) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: JDBCRDD will read metadata:
https://github.com/apache/spark/blob/v2.2.0-rc5/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala#L85
I'll change here next version.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 5]
```diff
@@ -82,7 +82,7 @@ object JDBCRDD extends Logging {
    * @return A Catalyst schema corresponding to columns in the given order.
    */
   private def pruneSchema(schema: StructType, columns: Array[String]): StructType = {
-    val fieldMap = Map(schema.fields.map(x => x.metadata.getString("name") -> x): _*)
+    val fieldMap = Map(schema.fields.map(x => x.name -> x): _*)
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: `x.metadata.getString("name")` always equals `x.name`:
https://github.com/apache/spark/blob/v2.2.0-rc5/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala#L291

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Then, we should remove this test case.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala [hunk 6]
```diff
@@ -120,6 +120,7 @@ class JDBCOptions(
   // TODO: to reuse the existing partition parameters for those partition specific options
   val createTableOptions = parameters.getOrElse(JDBC_CREATE_TABLE_OPTIONS, "")
   val createTableColumnTypes = parameters.get(JDBC_CREATE_TABLE_COLUMN_TYPES)
+  val customSchema = parameters.get(JDBC_CUSTOM_SCHEMA)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala]: convert it to `StructType` here.

**Comment by gatorsmile** [comment]
I am fine to support customized schema for read path of JDBC relation. However, we need to check whether the user-specified schema matches the underlying the table schema. If not matched, we need to capture it earlier and issue a proper error message.

**Review by leophan — APPROVED** [review]

**Comment by gatorsmile** [comment]
The example in the PR description looks a little bit confusing. 
```Scala
val dfRead = spark.read.schema(schema).jdbc(jdbcUrl, "tableWithCustomSchema", new Properties())
```

Could you update it?

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala [hunk 7]
```diff
@@ -268,4 +275,44 @@ class OracleIntegrationSuite extends DockerJDBCIntegrationSuite with SharedSQLCo
     assert(row.getDate(0).equals(dateVal))
     assert(row.getTimestamp(1).equals(timestampVal))
   }
+
+  test("SPARK-20427/SPARK-20921: read table use custom schema") {
+
+    // default will throw IllegalArgumentException
+    val e = intercept[org.apache.spark.SparkException] {
+      spark.read.jdbc(jdbcUrl, "custom_column_types", new Properties()).collect()
```
Comment (gatorsmile) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: Nit: Change the table names in all the test cases.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 8]
```diff
@@ -111,7 +111,22 @@ private[sql] case class JDBCRelation(
 
   override val needConversion: Boolean = false
 
-  override val schema: StructType = JDBCRDD.resolveTable(jdbcOptions)
+  override val schema: StructType = {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: add a comment to explain this is resolved by names.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 9]
```diff
@@ -111,7 +111,22 @@ private[sql] case class JDBCRelation(
...
+    val schema = JDBCRDD.resolveTable(jdbcOptions)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: this should not be called if the users already specify the schema

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 10]
```diff
@@ -111,7 +111,22 @@ private[sql] case class JDBCRelation(
...
+    val customSchema = jdbcOptions.customSchema
+    if (customSchema.isDefined) {
+      val schemaFieldNames = schema.fieldNames.mkString(",")
+      val customSchemaFieldNames = customSchema.get.fieldNames.mkString(",")
+      if(schemaFieldNames.equals(customSchemaFieldNames)) {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: We also need to consider the case sensitivity

**Comment by gatorsmile** [comment]
Users should be allowed to specify the schema from the table properties by using DDL-like strings.

**Comment by wangyum** [comment]
How about below? similar to [Hive JDBCStorageHandler](https://github.com/apache/hive/blob/rel/release-2.3.0/ql/src/test/queries/clientpositive/jdbc_handler.q#L1-L17):
```sql
CREATE TEMPORARY VIEW tableWithCustomSchema (
  ID decimal(38, 0),
  N1 int,
  N2 boolean
)
USING org.apache.spark.sql.jdbc
OPTIONS (url '$jdbcUrl', dbTable 'tableWithCustomSchema')
```

**Comment by gatorsmile** [comment]
So far, we do not want to introduce the user-specified schema to `CREATE TEMPORARY VIEW`, but we might do it in the other formats in the data source API v2.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 11]
```diff
@@ -268,7 +270,9 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
       table: String,
       predicates: Array[String],
       connectionProperties: Properties): DataFrame = {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Also need a test case for this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 12]
```diff
@@ -87,7 +87,11 @@ case class DataSource(
 
   lazy val providingClass: Class[_] = DataSource.lookupDataSource(className)
   lazy val sourceInfo: SourceInfo = sourceSchema()
-  private val caseInsensitiveOptions = CaseInsensitiveMap(options)
+  private val caseInsensitiveOptions = if (userSpecifiedSchema.isDefined) {
+    CaseInsensitiveMap(options + (JDBCOptions.JDBC_CUSTOM_SCHEMA -> userSpecifiedSchema.get.json))
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: Revert it back? This is a wrong place

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: This is not a related change. Could you revert it back?

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 13]
```diff
@@ -953,21 +953,6 @@ class JDBCSuite extends SparkFunSuite
     assert(new JDBCOptions(CaseInsensitiveMap(parameters)).asConnectionProperties.isEmpty)
   }
 
-  test("SPARK-16848: jdbc API throws an exception for user specified schema") {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Also add test cases in JDBCSuite.scala?

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala [hunk 14]
```diff
@@ -268,4 +275,64 @@ class OracleIntegrationSuite extends DockerJDBCIntegrationSuite with SharedSQLCo
     assert(row.getDate(0).equals(dateVal))
     assert(row.getTimestamp(1).equals(timestampVal))
   }
+
+  test("SPARK-20427/SPARK-20921: read table use custom schema by jdbc api") {
```
Comment (gatorsmile) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: Move these newly added test cases to JDBCSuite.scala. 

Only add a single test case for Oracle data type mappings requested in the JIRA tickets.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 15]
```diff
@@ -111,7 +111,23 @@ private[sql] case class JDBCRelation(
...
+    val schema = JDBCRDD.resolveTable(jdbcOptions)
+    val customSchema = jdbcOptions.customSchema
+    if (customSchema.isDefined) {
+      val fieldNames = schema.fieldNames.mkString(",")
+      val customFieldNames = customSchema.get.fieldNames.mkString(",")
+      if ((sqlContext.conf.caseSensitiveAnalysis && fieldNames.equals(customFieldNames)) ||
+        fieldNames.toLowerCase.equals(customFieldNames.toLowerCase)) {
+        customSchema.get
+      } else {
+        throw new IllegalArgumentException(s"Field $customFieldNames does not match $fieldNames.")
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: Ask users to provide all the columns.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 16]
```diff
@@ -111,7 +111,23 @@ private[sql] case class JDBCRelation(
...
 
-  override val schema: StructType = JDBCRDD.resolveTable(jdbcOptions)
+  // This is resolved by names, only check the column names.
+  override val schema: StructType = {
+    val schema = JDBCRDD.resolveTable(jdbcOptions)
+    val customSchema = jdbcOptions.customSchema
+    if (customSchema.isDefined) {
+      val fieldNames = schema.fieldNames.mkString(",")
+      val customFieldNames = customSchema.get.fieldNames.mkString(",")
+      if ((sqlContext.conf.caseSensitiveAnalysis && fieldNames.equals(customFieldNames)) ||
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: Please follow what we did in `parseUserSpecifiedCreateTableColumnTypes`

**Comment by wangyum** [comment]
Jenkins, retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 17]
```diff
@@ -768,6 +769,25 @@ object JdbcUtils extends Logging {
...
+    val userSchema = CatalystSqlParser.parseTableSchema(columnTypes)
+    // This is resolved by names, only check the column names.
+    userSchema.fieldNames.foreach { col =>
+      schema.find(f => nameEquality(f.name, col)).getOrElse {
+        throw new AnalysisException(
+          s"${JDBCOptions.JDBC_CUSTOM_DATAFRAME_COLUMN_TYPES} option column $col not found in " +
+            s"schema ${schema.catalogString}")
+      }
+    }
+    userSchema
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: What is your expected behaviors when users-specified schema does not include all the columns of the underlying table schema?

File: examples/src/main/python/sql/datasource.py [hunk 18]
```diff
@@ -177,6 +177,16 @@ def jdbc_dataset_example(spark):
...
               properties={"user": "username", "password": "password"})
 
+    # Specifying dataframe column data types on read
+    jdbcDF3 = spark.read \
+        .format("jdbc") \
+        .option("url", "jdbc:postgresql:dbserver") \
+        .option("dbtable", "schema.tablename") \
+        .option("user", "username") \
+        .option("password", "password") \
+        .option("customDataFrameColumnTypes", "id decimal(38, 0), name string") \
```
Comment (gatorsmile) [inline examples/src/main/python/sql/datasource.py]: `decimal ` -> `DECIMAL`
`string` -> `STRING`

File: examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala [hunk 19]
```diff
@@ -185,6 +185,10 @@ object SQLDataSourceExample {
     connectionProperties.put("password", "password")
     val jdbcDF2 = spark.read
       .jdbc("jdbc:postgresql:dbserver", "schema.tablename", connectionProperties)
+    // Specifying dataframe column data types on read
+    connectionProperties.put("customDataFrameColumnTypes", "id decimal(38, 0), name string")
```
Comment (gatorsmile) [inline examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala]: The same here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 5 above)
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: [`CatalystSqlParser.parseTableSchema(columnTypes)`](https://github.com/apache/spark/blob/v2.2.0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParseDriver.scala#L63) constructed StructType without metadata, error message:
```
key not found: name
java.util.NoSuchElementException: key not found: name
	at scala.collection.MapLike$class.default(MapLike.scala:228)
	at scala.collection.AbstractMap.default(Map.scala:59)
	at scala.collection.MapLike$class.apply(MapLike.scala:141)
	at scala.collection.AbstractMap.apply(Map.scala:59)
	at org.apache.spark.sql.types.Metadata.get(Metadata.scala:111)
	at org.apache.spark.sql.types.Metadata.getString(Metadata.scala:60)
	at org.apache.spark.sql.execution.datasources.jdbc.JDBCRDD$$anonfun$1.apply(JDBCRDD.scala:83)
	at org.apache.spark.sql.execution.datasources.jdbc.JDBCRDD$$anonfun$1.apply(JDBCRDD.scala:83)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 17 above)
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: OK, I'll ask user to specified all the columns.

**Comment by sobusiak** [comment]
As far as I understand the proposed solution recommends using `DecimalType` with user-chosen precision and scale but `DecimalType` cannot represent all numbers that can be stored in NUMBER. I believe `DoubleType` is more suitable. Actually why isn't it the default Spark type for Oracle's NUMBER? Please see my comment in SPARK-20427.

**Comment by wangyum** [comment]
Yes, mapping to Double seems fine. this test passed:
```
  test("SPARK-20427/SPARK-20921: read table use custom schema by jdbc api") {
    // default will throw IllegalArgumentException
    val e = intercept[org.apache.spark.SparkException] {
      spark.read.jdbc(jdbcUrl, "tableWithCustomSchema", new Properties()).collect()
    }
    assert(e.getMessage.contains(
      "requirement failed: Decimal precision 39 exceeds max precision 38"))

    // custom schema can read data
    val props = new Properties()
    props.put("customDataFrameColumnTypes",
      s"ID double, N1 int, N2 boolean")
    val dfRead = spark.read.jdbc(jdbcUrl, "tableWithCustomSchema", props)

    val rows = dfRead.collect()
    // verify the data type
    val types = rows(0).toSeq.map(x => x.getClass.toString)
    assert(types(0).equals("class java.lang.Double"))
    assert(types(1).equals("class java.lang.Integer"))
    assert(types(2).equals("class java.lang.Boolean"))

    // verify the value
    val values = rows(0)
    assert(values.getDouble(0).equals(12312321321321312312312312123D))
    assert(values.getInt(1).equals(1))
    assert(values.getBoolean(2).equals(false))
  }
```

**Comment by gatorsmile** [comment]
will review this today.

File: docs/sql-programming-guide.md [hunk 20]
```diff
@@ -1334,7 +1334,14 @@ the following case-insensitive options:
     <td>
      The database column data types to use instead of the defaults, when creating the table. Data type information should be specified in the same format as CREATE TABLE columns syntax (e.g: <code>"name CHAR(64), comments VARCHAR(1024)")</code>. The specified types should be valid spark sql data types. This option applies only to writing.
     </td>
-  </tr>  
+  </tr>
+
+  <tr>
+    <td><code>customDataFrameColumnTypes</code></td>
```
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: `readTableColumnTypes`

File: docs/sql-programming-guide.md [hunk 21]
```diff
@@ -1334,7 +1334,14 @@ the following case-insensitive options:
...
+    <td>
+     The DataFrame column data types to use instead of the defaults when reading data from jdbc API. (e.g: <code>"id DECIMAL(38, 0), name STRING")</code>. The specified types should be valid spark sql data types. This option applies only to reading.
```
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: This is not limited to DataFrame. 

> The customized column types to use for reading data from JDBC connectors. For example, <code>"id DECIMAL(38, 0), name STRING")</code>. The specified types should be valid spark sql data types. This option applies only to reading.

File: examples/src/main/python/sql/datasource.py [hunk 22]
```diff
@@ -177,6 +177,16 @@ def jdbc_dataset_example(spark):
...
               properties={"user": "username", "password": "password"})
 
+    # Specifying dataframe column data types on read
+    jdbcDF3 = spark.read \
+        .format("jdbc") \
+        .option("url", "jdbc:postgresql:dbserver") \
+        .option("dbtable", "schema.tablename") \
+        .option("user", "username") \
+        .option("password", "password") \
+        .option("customDataFrameColumnTypes", "id DECIMAL(38, 0), name STRING") \
```
Comment (gatorsmile) [inline examples/src/main/python/sql/datasource.py]: `readTableColumnTypes`

File: examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala [hunk 23]
```diff
@@ -185,6 +185,10 @@ object SQLDataSourceExample {
     connectionProperties.put("password", "password")
     val jdbcDF2 = spark.read
       .jdbc("jdbc:postgresql:dbserver", "schema.tablename", connectionProperties)
+    // Specifying dataframe column data types on read
+    connectionProperties.put("customDataFrameColumnTypes", "id DECIMAL(38, 0), name STRING")
```
Comment (gatorsmile) [inline examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala]: `customSchema `

File: examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala [hunk 24]
```diff
@@ -185,6 +185,10 @@ object SQLDataSourceExample {
     connectionProperties.put("password", "password")
     val jdbcDF2 = spark.read
       .jdbc("jdbc:postgresql:dbserver", "schema.tablename", connectionProperties)
+    // Specifying dataframe column data types on read
```
Comment (gatorsmile) [inline examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala]: > Specifying the custom data types of columns

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 25]
```diff
@@ -679,6 +679,16 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
   }
 
   /**
+   * A convenient function for validate specified column types schema in jdbc API.
+   */
+  private def assertJdbcAPISpecifiedDataFrameSchema(): Unit = {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: `assertJdbcAPISpecifiedDataFrameSchema` -> `assertNoSpecifiedSchemaForJDBC`

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 26]
```diff
@@ -197,7 +197,7 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
    * @since 1.4.0
    */
   def jdbc(url: String, table: String, properties: Properties): DataFrame = {
-    assertNoSpecifiedSchema("jdbc")
+    assertJdbcAPISpecifiedDataFrameSchema()
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Users should be able to do it in either way. If users specify them in both `schema()` API and the `customerSchema` option, we should issue an exception.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 27]
```diff
@@ -80,7 +80,7 @@ object JDBCRDD extends Logging {
...
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Sorry, I did not get your point. Could you show me an example? Is it a behavior breaking change?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 28]
```diff
@@ -111,7 +111,16 @@ private[sql] case class JDBCRelation(
 
   override val needConversion: Boolean = false
 
-  override val schema: StructType = JDBCRDD.resolveTable(jdbcOptions)
+  override val schema: StructType = {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: Please address the above comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 27 above)
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]:
```
scala> org.apache.spark.sql.catalyst.parser.CatalystSqlParser.parseTableSchema("id int, name string").fields.map(x => x.metadata.getString("name") -> x)
java.util.NoSuchElementException: key not found: name
  at scala.collection.MapLike$class.default(MapLike.scala:228)
  at scala.collection.AbstractMap.default(Map.scala:59)
  at scala.collection.MapLike$class.apply(MapLike.scala:141)
  at scala.collection.AbstractMap.apply(Map.scala:59)
  at org.apache.spark.sql.types.Metadata.get(Metadata.scala:111)
  at org.apache.spark.sql.types.Metadata.getString(Metadata.scala:60)
  at $anonfun$1.apply(<console>:24)
  at $anonfun$1.apply(<console>:24)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
  at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
  at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
  at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
  at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
  ... 48 elided
```

**Review by gatorsmile — COMMENTED** [review]
Looks pretty good! Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 27 above)
[the remark this answers is not in the document]
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: It seems safe to remove this line.

**Comment by gatorsmile** [comment]
@wangyum Could you update the example in the PR description?

**Comment by wangyum** [comment]
@gatorsmile Done

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Pull request closed without merging**